src/servers/src/grpc/flight.rs

```diff
@@ -98,24 +97,28 @@ impl FlightService for FlightHandler {
         })?;
         let query_ctx = create_query_context(request.header.as_ref());
 
-        let (tx, rx) = oneshot::channel();
         let handler = self.handler.clone();
 
         // Executes requests in another runtime to
         // 1. prevent the execution from being cancelled unexpected by Tonic runtime;
+        //   - Refer to our blog for the rational behind it:
+        //     https://www.greptime.com/blogs/2023-01-12-hidden-control-flow.html
+        //   - Obtaining a `JoinHandle` to get the panic message (if there's any).
+        //     From its docs, `JoinHandle` is cancel safe. The task keeps running even it's handle been dropped.
         // 2. avoid the handler blocks the gRPC runtime incidentally.
-        self.runtime.spawn(async move {
-            let result = handler.do_query(query, query_ctx).await;
-
-            // Ignore the sending result.
-            // Usually an error indicates the rx at Tonic side is dropped (due to request timeout).
-            let _ = tx.send(result);
-        });
-
-        // Safety: An early-dropped tx usually indicates a serious problem (like panic).
-        // This unwrap is used to poison the upper layer.
-        let output = rx.await.unwrap()?;
-
+        let handle = self
+            .runtime
+            .spawn(async move { handler.do_query(query, query_ctx).await });
+
+        let output = handle.await.map_err(|e| {
+            if e.is_cancelled() {
+                Status::cancelled(e.to_string())
```

Comment:
I'm afraid this branch doesn't work as expected. When cancellation happens it starts from the upper level, rather than inside the task. It's an unreachable branch in our current implementation.

Reply:
Yes, the cancellation here is not triggered by Tonic. Instead, it's the runtime that handles the task cancel it. In current implementation, the task is indeed not being cancelled. However, I think we will very likely implement a cancellable runtime in the future, and checking `if e.is_cancelled()` here is the proper way to handle the cancellation.

It might be naive to handle cancellation like this, but we will find a better way when actually implement it.